sys/evtimer/evtimer.c

```diff
@@ -31,64 +31,53 @@
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
-/* XXX this function is intentionally non-static, since the optimizer can't
- * handle the pointer hack in this function */
-void evtimer_add_event_to_list(evtimer_t *evtimer, evtimer_event_t *event)
+static void evtimer_add_event_to_list(evtimer_t *evtimer, evtimer_event_t *event)
 {
     DEBUG("evtimer: new event offset %" PRIu32 " ms\n", event->offset);
-    /* we want list->next to point to the first list element. thus we take the
-     * *address* of evtimer->events, then cast it from (evtimer_event_t **) to
-     * (evtimer_event_t*). After that, list->next actually equals
-     * evtimer->events. */
-    evtimer_event_t *list = (evtimer_event_t *)&evtimer->events;
-
-    while (list->next) {
-        evtimer_event_t *list_entry = list->next;
+    evtimer_event_t **list = &evtimer->events;
```

Comment:
Why not `evtimer_event_t *list = evtimer->events;` and replace `*list` by `list` in the following lines?

Reply:
Well, it is quite essential to have `evtimer_event_t **list = &evtimer->events` !!
Take a look at the last line in the function
```
    *list = event;
```

Reply:
Why not using `evtime_event_t *list` and stopping when `list->next` `event` is to be inserted between `list` and `list->next`? Than you would update `event->next` to point to `list->next` and update `list->next` to point to `event`. This would make the whole code much more readable.

Reply:
:-) I like you're persistence to want `evtime_event_t *list`.

Have you thought about inserting the first event, when `evtimer->events` is NULL? I think not.

Again I want to emphasize that having the pointer pointer is there on purpose. First it is used to find the place where the list ends (including the situation of an empty list). And when it finds it it already points to the place where the element must be inserted.

Reply:
Let me explain why I think it is beneficial to avoid an extra layer of indirection if this can be done without pain: As part of my job I teach low level programming (among other stuff) to computer science students. From my experience at most every second student is able to understand the concept of pointers, as long as you remain at one level of indirection (pointer to foo). When you add one more layer (pointer to pointer to foo), at most one out of ten students will be able to understand the code.

While this is basically anecdotal evidence, I'm pretty sure that using 2 levels of indirection (pointer to pointer to foo) will give a lot of people a hard time trying to understand the code. There clearly are places where using pointers to pointers can make the code much shorter, more performant, or may just not be practical to implement with 1 level of indirection. But when there is no "pain" in using pointer to foo instead of pointer to pointer to foo, this will make a lot of peoples life easier ;-)

Let me sketch how this could be done with 1 level of indirection. (Untested code without having spend more than a second of thought.):

``` C
static void evtimer_add_event_to_list(evtimer_t *evtimer, evtimer_event_t *event)
{
    if (!evtimer->events || (event->offset < evtimer->events->offset)) {
        /* insert as first element */
        event->next = evtimer->events;
        evtimer->events = event;
    }
    else {
        evtimer_event_t *i;
        for (i = evtimer->events; i->next; i = i->next) {
            /* Set event offset relative to previous event */
            event->offset -= i->offset;

            if (event->offset < i->next->offset) {
                break;
            }
        }
        /* insert event between i and i->next */
        event->next = i->next;
        i->next = event;
    }

    if (event->next) {
        /* Updating offset of the follow-up to the newly inserted event to
           be relative to the newly inserted one */
        event->next->offset -= event->offset;
    }
}
```

Reply:
I'm in favor of easy to understand code. But I still prefer my version. What else can I do to convince you? Shall we measure code complexity (McCabe?), or shall we measure code size? Do we need more comments (it's basically a simple while-loop), explaining the iteration. ASCII art?

Leaving out DEBUG, it comes down to the following.
```
static void evtimer_add_event_to_list(evtimer_t *evtimer, evtimer_event_t *event)
{
    evtimer_event_t **list = &evtimer->events;

    while (*list) {
        /* Stop when new event time is nearer then next */
        if (event->offset < (*list)->offset) {
            break;
        }
        /* Set event offset relative to previous event */
        event->offset -= (*list)->offset;
        list = &(*list)->next;
    }

    /* Set found next bigger event after new event */
    event->next = *list;
    if (*list) {
        /* Set offset following event relative to new event */
        (*list)->offset -= event->offset;
    }
    *list = event;
}
```

Reply:
Note that I deliberately did not mark this comment as "changes requested".

But out of curiosity I did compare the assembly emitted from GCC and clang. Your code and the current master are translated to the exact same assembly with clang8 on x86 with my suggestion being a bit longer. With GCC 8.3.0 on ARM your code is translated to the shortest assembly, followed by the master code and my code being the longest. With GCC 8.3.0 on x86 your code translates again to the shortest assembly, followed by my code with the current master being the longest.